src/librekey/rnp_key_store.cpp

```diff
@@ -588,43 +588,97 @@ rnp_key_store_get_signer_key(rnp_key_store_t *store, const pgp_signature_t *sig)
     return NULL;
 }
 
+static pgp_sig_import_status_t
+rnp_key_store_import_subkey_signature(rnp_key_store_t *      keyring,
+                                      pgp_key_t *            key,
+                                      const pgp_signature_t *sig)
+{
+    pgp_sig_type_t sigtype = signature_get_type(sig);
+    if ((sigtype != PGP_SIG_SUBKEY) && (sigtype != PGP_SIG_REV_SUBKEY)) {
+        return PGP_SIG_IMPORT_STATUS_UNKNOWN;
+    }
+    const uint8_t *prim_grip = pgp_key_get_primary_grip(key);
+    pgp_key_t *    primary = rnp_key_store_get_signer_key(keyring, sig);
+    if (!prim_grip || !primary) {
+        RNP_LOG("No primary grip or primary key");
+        return PGP_SIG_IMPORT_STATUS_UNKNOWN_KEY;
+    }
+    if (memcmp(pgp_key_get_grip(primary), prim_grip, PGP_KEY_GRIP_SIZE)) {
+        RNP_LOG("Wrong subkey signature's signer.");
+        return PGP_SIG_IMPORT_STATUS_UNKNOWN;
+    }
+
+    pgp_key_t tmpkey = {};
+    if (!pgp_key_from_pkt(&tmpkey, &key->pkt) || !rnp_key_add_signature(&tmpkey, sig)) {
+        RNP_LOG("Failed to add signature to the key.");
+        pgp_key_free_data(&tmpkey);
+        return PGP_SIG_IMPORT_STATUS_UNKNOWN;
+    }
+
+    size_t expackets = pgp_key_get_rawpacket_count(key);
+    if (!(key = rnp_key_store_add_key(keyring, &tmpkey))) {
+        RNP_LOG("Failed to add key with imported sig to the keyring");
+        pgp_key_free_data(&tmpkey);
+        return PGP_SIG_IMPORT_STATUS_UNKNOWN;
+    }
+    return (pgp_key_get_rawpacket_count(key) > expackets) ? PGP_SIG_IMPORT_STATUS_NEW :
+                                                            PGP_SIG_IMPORT_STATUS_UNCHANGED;
+}
+
+pgp_sig_import_status_t
+rnp_key_store_import_key_signature(rnp_key_store_t *      keyring,
+                                   pgp_key_t *            key,
+                                   const pgp_signature_t *sig)
+{
+    if (pgp_key_is_subkey(key)) {
+        return rnp_key_store_import_subkey_signature(keyring, key, sig);
+    }
+    pgp_sig_type_t sigtype = signature_get_type(sig);
+    if ((sigtype != PGP_SIG_DIRECT) && (sigtype != PGP_SIG_REV_KEY)) {
+        RNP_LOG("Wrong signature type: %d", (int) sigtype);
+        return PGP_SIG_IMPORT_STATUS_UNKNOWN;
+    }
+
+    pgp_key_t tmpkey = {};
+    if (!pgp_key_from_pkt(&tmpkey, &key->pkt) || !rnp_key_add_signature(&tmpkey, sig)) {
+        RNP_LOG("Failed to add signature to the key.");
+        pgp_key_free_data(&tmpkey);
+        return PGP_SIG_IMPORT_STATUS_UNKNOWN;
+    }
+
+    size_t expackets = pgp_key_get_rawpacket_count(key);
+    if (!(key = rnp_key_store_add_key(keyring, &tmpkey))) {
```

Comment:
@ni4 I noticed that `rnp_key_store_add_key` actually has no effect (doesn't update) on `srckey->format == PGP_KEY_STORE_G10`. So I guess unit tests will fail with this keystore format. Is this behaviour desired?

Reply:
@rrrooommmaaa pgp_key_from_pkt() sets format to PGP_KEY_STORE_GPG, so it will not fail in this case.